### PR TITLE
[API] As a user, I can view past report

### LINKED ISSUE #24: [API] As a user, I can view past report
### Why?
User can make an API request contain with id of the stored data to view the report in the past

### Acceptance Criteria
- [x] Provide an endpoint to view report
- [x] Return report data

### PR DESCRIPTION
**Pull request by Lahphim**
Resolve https://github.com/Lahphim/go-crawler-challenge/issues/24

## What happened 👀

The user can make a request to get the report detail from this endpoint `/api/v1/report/:keyword_id`.
In case that the report haven't completed with the service scrapping google page, it will not report the report detail 😛

## Insight 📝

- [x] Provide an endpoint to view report
- [x] Return report data

Sample request:
```curl
curl --location --request GET 'http://localhost:8080/api/v1/report/1' \
--header 'Authorization: Bearer Y2MZOTG2NWITZGZIMY0ZN2Q3LTHKNJGTZJIZODZLNZDIMZM4'
```

## Proof Of Work 📹

[GET] /api/v1/report/:keyword_id
Show report detail

| Valid Request | Invalid Request |
|---------------|----------------|
| Screen Shot 2564-04-05 at 17 21 05 | Screen Shot 2564-04-05 at 17 21 20 |

### CODE REVIEW

File: controllers/api/v1/report.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	keywordId := c.GetString(":keyword_id")
+	query := map[string]interface{}{
+		"id":      keywordId,
+		"user_id": c.CurrentUser.Id,
+		"status":  models.GetKeywordStatus("completed"),
+	}
+
+	keyword, err := models.GetKeywordBy(query, []string{})
+	if err != nil {
+		c.RenderGenericError(ErrorNotFoundReport)
```
Comment (hoangmirs) [inline controllers/api/v1/report.go]: Should we return `404 NOTFOUND` in this case instead?

File: controllers/api/v1/report_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		TruncateTable("link")
+		TruncateTable("page")
+		TruncateTable("position")
+		TruncateTable("keyword")
+		TruncateTable("user")
+	})
+
+	Describe("GET /api/v1/report/:keyword_id", func() {
+		Context("given a valid access token", func() {
+			Context("given a valid report", func() {
```
Comment (hoangmirs) [inline controllers/api/v1/report_test.go]: Should it be "given a valid keyword ID" as we're requesting the report of a keyword?

File: controllers/api/v1/report_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+					headers := http.Header{"Authorization": {fmt.Sprintf("Bearer %v", accessToken)}}
+
+					response := MakeAuthenticatedRequest("GET", fmt.Sprintf("/api/v1/report/%v", keyword.Id), headers, nil, nil)
+
+					Expect(response).To(MatchJSONSchema("report/show/valid"))
+				})
+			})
+
+			Context("given an INVALID report", func() {
+				It("returns status 500", func() {
```
Comment (hoangmirs) [inline controllers/api/v1/report_test.go]: As I said above, should it be `404` instead?

**Review by hoangmirs — APPROVED** [review]

File: controllers/api/v1/report.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+func (c *ReportController) actionPolicyMapping() {
+	c.MappingPolicy("Show", Policy{RequireAuthenticatedUser: true})
+}
+
+// Show handles report detail
+// @Title Show
+// @Description show the search result of the given keyword stored in the database
+// @Success 202 {object} v1serializers.ReportDetail
+// @Param :keyword_id path string true
+// @Failure 500 Internal Server Error
```
Comment (hoangmirs) [inline controllers/api/v1/report.go:38]: We should edit this as well

File: serializers/v1/report_detail.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+	"go-crawler-challenge/models"
+)
+
+type ReportDetail struct {
+	Report *models.Report
+}
+
+type reportDetailResponse struct {
+	Id      string `jsonapi:"primary,reports"`
```
Comment (junan) [inline serializers/v1/report_detail.go]: Should the type be `singular`(`report`) as it is fetching single report?

**Review by junan — APPROVED** [review]

**Pull request merged**